### PR TITLE
Support URLs such as https://domain/@scope/package

### LINKED ISSUE #6: unpkg (et al?) must have non-zero directory
Currently udd does this which is wrong:

```diff
-// https://unpkg.com/@types/jszip@3.1.7/index.d.ts
+// https://unpkg.com/@undefined/jszip@3.1.7/index.d.ts
```

### PR DESCRIPTION
**Pull request by takker99**
close #6

### CODE REVIEW

File: registry.ts [hunk 1]
```diff
@@ -239,6 +298,40 @@ export class Denopkg implements RegistryUrl {
...
+  at(version: string): RegistryUrl {
+    const { parts, packageName } = this.parts();
+    parts[4] = `${packageName}@${version}`;
+    return new PikaScope(parts.join("/"));
+  }
+
+  version(): string {
+    const { version } = this.parts();
+    if (version === undefined) {
+      throw Error(`Unable to find version in ${this.url}`);
```
Comment (hayd) [inline registry.ts:326]: won't this be thrown already in defaultParts? https://github.com/hayd/deno-udd/pull/53/files#diff-58f42611a053bb45add25a11c10ed01c89fa3d17a917ecea2de4fad160568ff8R173

File: registry.ts [hunk 2]
```diff
@@ -154,6 +154,65 @@ async function unpkgVersions(name: string): Promise<string[]> {
...
+
+  async all(): Promise<string[]> {
+    const { scope, packageName } = this.parts();
+    return await unpkgVersions(`${scope}/${packageName}`);
+  }
+
+  at(version: string): RegistryUrl {
+    const { parts, packageName } = this.parts();
+    parts[4] = `${packageName}@${version}`;
+    return new UnpkgScope(parts.join("/"));
```
Comment (hayd) [inline registry.ts:202]: To decrease some boilerplate perhaps:
```
function defaultScopeAt(that: RegistryUrl): string {
   const { parts, packageName } = defaultParts(this);
   parts[4] = `${packageName}@${version}`;
   return parts.join("/");
}
```
could be pulled out or something?

**Comment by hayd** [comment]
Wow, this is great. Slightly agree with your comment about boilerplate but this functionality is great.
So I am happy to merge if you are.

(That said, it might be good to have some erroring cases for the parser - i.e. things that throw. But.. looking at test file there aren't any for any of the REGISTRIES! 😳 )

File: registry.ts (hunk 1 above)
Comment (takker99) [reply registry.ts:326]: Oh, that's right. I should have paid more attention.
Thank you for your pointing. I'll fix it.

File: registry.ts (hunk 2 above)
Comment (takker99) [reply registry.ts:202]: Looks good!
I consider whether to merge `defaultScopeAt` into `defaultAt` like this:

```ts
export function defaultAt(
  that: RegistryUrl,
  version: string,
  hasScope = false,
): string {
  if (!hasScope) return that.url.replace(/@(.*?)(\/|$)/, `@${version}/`);

  const { parts, packageName } = defaultParts(that);
  parts[4] = `${packageName}@${version}`;
  return parts.join("/");
}
```

but this refactoring merely makes registry.ts little shorter, so I'll use the code you suggested.

**Comment by takker99** [comment]
I'm sorry for late reply. I appreciate your review!
Would you review again and merge the PR?

File: registry.ts [hunk 3]
```diff
@@ -154,6 +154,76 @@ async function unpkgVersions(name: string): Promise<string[]> {
   return m.map((x) => x[1]);
 }
 
+interface PackageInfo {
+  parts: string[];
+  scope: string;
+  packageName: string;
+  version: string;
+}
+function defaultParts(that: RegistryUrl): PackageInfo {
```
Comment (hayd) [inline registry.ts:163]: nit: this should probably have been defaultInfo etc ? 😆 
but it's fine.

File: registry.ts [hunk 4]
```diff
@@ -154,6 +154,76 @@ async function unpkgVersions(name: string): Promise<string[]> {
...
+  const { parts, packageName } = defaultParts(that);
+  parts[4] = `${packageName}@${version}`;
+  return parts.join("/");
+}
+export function _defaultAt(
+  that: RegistryUrl,
+  version: string,
+  hasScope = false,
+): string {
+  if (!hasScope) return that.url.replace(/@(.*?)(\/|$)/, `@${version}/`);
```
Comment (hayd) [inline registry.ts:192]: This is probably too clever :), can you delete _defaultAt  (unused I think) ?

**Comment by hayd** [comment]
Thanks @takker99 , If you've tested this on a project or two I am happy to merge! 💪

File: registry.ts (hunk 4 above)
Comment (takker99) [reply registry.ts:192]: Oops! I forgot deleting this. Thanks!
(I wonder if there is such a lint feature which tells me unused export functions and values...)

**Comment by takker99** [comment]
Thanks! I've fixed them!

**Comment by hayd** [comment]
Fantastic! Thanks for this @takker99 , this is a very valuable feature.

**Pull request merged**